src/nvim/buffer.c

```diff
@@ -309,7 +309,7 @@ close_buffer (
   // buffers.
   // The caller must take care of NOT deleting/freeing when 'bufhidden' is
   // "hide" (otherwise we could never free or delete a buffer).
-  if (!buf->terminal) {
+  if (!buf->terminal && !buf->b_messages) {
```

Comment:
The main thing I was going for by following the terminal checks is I didn't want the message buffer to be editable at all.  I think that if a user could change certain buffer settings, it could lead to some bad behavior from writing messages to a buffer that the user expects to be controlling.